File: 02946. Matrix Similarity After Cyclic Shifts/solution.cpp
/**
 * Problem: 2946. Matrix Similarity After Cyclic Shifts
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution {
public:
  bool areSimilar(vector<vector<int>> &mat, int k) {
    // Get the number of rows and columns in the matrix
    int rows = mat.size(), cols = mat[0].size();

    // Check each row in the matrix
    for (int row = 0; row < rows; row++) {
      // Check each element in current row
      for (int col = 0; col < cols; col++) {
        // Calculate the shifted column index
        int shiftedCol = (col + k) % cols;

        // Compare the current element with the shifted element
        if (mat[row][col] != mat[row][shiftedCol])
          return false;
      }
    }

    // If all elements are equal to their shifted counterparts, return true
    return true;
  }
};

File: 02946. Matrix Similarity After Cyclic Shifts/solution.ts
/**
 * Problem: 2946. Matrix Similarity After Cyclic Shifts
 *
 * Difficulty: Easy
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Checks if matrix remains the same after shifting each row left by k positions
 *
 * @param mat - Input matrix
 * @param k - Number of positions to shift left
 *
 * @returns True if matrix is similar after shifting
 */
const areSimilar = (mat: number[][], k: number): boolean =>
  // Check each row in the matrix
  mat.every((row: number[], rowIndex: number) =>
    // Check each element in current row
    row.every(
      (value: number, colIndex: number) =>
        // Compare with element at shifted position
        // Formula: original column + k (mod total columns)
        value === mat[rowIndex][(colIndex + k) % mat[0].length]
    )
  )
